design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+
+## Goals
+
+To solve the problems with ECS construct implemention we propose adding a new `Service` class with the following goals:
+
+- __One service class to learn__: Instead of a collection of many different L3 classes that all have different names there is a single class which can implement any number of different forms depending on what the customer wants. Customers don't need to learn the names for different L3 constructs. All architectures start the same way: with the same `Service` class.
+- __Easy to extend and customize__: It is easy to extend the `Service` class with new behaviors and features, in any combination required. The `Service` class comes with batteries included: a set of easy to use extensions created and maintained by the AWS team. However there are also clear instructions and examples of how customers can create their own custom extensions. Third parties will be enabled to also create and distribute extensions for their own services, for example a provider like DataDog could easily build an extension that automatically adds DataDog agent to an ECS service.
+- __Extensions self configure and self provision__: Service extensions provision their own resources as well as configuring the ECS service to use those resources. For example a load balancer addon creates its own load balancer and also attaches the ECS service to the load balancer. An App Mesh extension creates its own App Mesh virtual service and virtual node as well as configuring the task definition to have the right Envoy sidecar and proxy settings.
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a service extension. Extensions know of each other's existence. For example the application's container is aware that there is an Envoy proxy and that it must wait until the Envoy proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing up the capacity strategy in the class name customers just create a service, and add it to an environment. The service automatically modifies its settings as needed to adjust to the capacity available in the environment.
+
+
+## Implementation
+
+The new `Service` class acheives its goals via a method `Service.add()` This method is used to add extensions to the service. Each extension is responsible for customizing the service to support a specific feature. The following example shows how the end user instantiates a service with various feature's enabled:
+
+```js
+const myService = new Service(stack, 'my-service', {});
+myService.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myService.add(new AppMeshAddon({ mesh }));
+myService.add(new FireLensAddon());
+myService.add(new XRayAddon());
+myService.add(new CloudwatchAgentAddon());
+myService.add(new HttpLoadBalancerAddon());
+```
```

Comment:
I love how streamlined this API is, and how we encapsulate extension logic into their own classes. However, do you anticipate certain extensions conflicting with each other? Given that ECS patterns are meant to push best practices in application infrastructure/architecture, how will we provide that guidance with extensions?

Reply:
Adding these two questions to the doc for discussion